object/thread.go

```diff
@@ -0,0 +1,98 @@
+package object
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/risor-io/risor/op"
+)
+
+type Thread struct {
+	callable Callable
+	args     []Object
+	done     chan bool
+	result   Object
+}
+
+func (t *Thread) Type() Type {
+	return THREAD
+}
+
+func (t *Thread) Inspect() string {
+	switch obj := t.callable.(type) {
+	case Object:
+		return fmt.Sprintf("thread(%s)", obj.Inspect())
+	default:
+		return "thread()"
+	}
+}
+
+func (t *Thread) Interface() interface{} {
+	return nil
+}
+
+func (t *Thread) IsTruthy() bool {
+	return true
+}
+
+func (t *Thread) Cost() int {
+	return 8
+}
+
+func (t *Thread) MarshalJSON() ([]byte, error) {
+	return nil, fmt.Errorf("type error: unable to marshal %s", THREAD)
+}
+
+func (t *Thread) RunOperation(opType op.BinaryOpType, right Object) Object {
+	return Errorf("eval error: unsupported operation for %s: %v", THREAD, opType)
+}
+
+func (t *Thread) Equals(other Object) Object {
+	return NewBool(t == other)
+}
+
+func (t *Thread) SetAttr(name string, value Object) error {
+	return fmt.Errorf("attribute error: %s object has no attribute %q", THREAD, name)
+}
+
+func (t *Thread) GetAttr(name string) (Object, bool) {
+	switch name {
+	case "wait":
+		return NewBuiltin("thread.wait", func(ctx context.Context, args ...Object) Object {
+			// Wait for the thread to finish or the context to be cancelled
+			select {
+			case <-ctx.Done():
+				return Errorf("eval error: %s", ctx.Err())
+			case <-t.done:
+				return t.result
+			}
+		}), true
+	}
+	return nil, false
+}
+
+func NewThread(ctx context.Context, callable Callable, args []Object) *Thread {
+	if callable == nil {
+		panic("callable is nil")
+	}
+
+	t := &Thread{
+		callable: callable,
+		args:     args,
+		done:     make(chan bool),
+	}
+
+	ctx = WithThread(ctx, t)
+
+	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				t.result = NewError(fmt.Errorf("panic: %v", r))
+			}
+			close(t.done)
+		}()
+		t.result = callable.Call(ctx, args...)
+	}()
```

Comment:
So this is doing real parallel execution. It is interesting, but I'm worried it forces Risor to also expose the other sync utilities like `sync.Mutex` and concurrent containers like `sync.Map`.

Maybe that's not horrible? Or maybe it'll be enough to rely on channels?

I'm just weary about the next steps after this PR.

Reply:
I think leveraging Go and its goroutines under the hood is the right move. Trying to do concurrency within the Risor VM in another way is going to be significantly more work and less performant.

I agree it's a slippery slope to bringing in other sync utils, but that's not a bad situation really. We can do it when and where it makes sense. Adding a Mutex type to Risor soon seems easy enough... at least that's a small API.

Reply:
As long as you're fine with it.